app/assets/javascripts/discourse/app/lib/sticky-avatars.js

```diff
@@ -0,0 +1,104 @@
+import { addWidgetCleanCallback } from "discourse/components/mount-widget";
+import Site from "discourse/models/site";
+import { bind } from "discourse-common/utils/decorators";
+import { schedule } from "@ember/runloop";
+
+export default class StickyAvatars {
+  stickyClass = "sticky-avatar";
+  topicPostSelector = "#topic .post-stream .topic-post";
+  intersectionObserver = null;
+  direction = "⬇️";
+  prevOffset = -1;
+
+  static init(container) {
+    new this(container).init();
+  }
+
+  constructor(container) {
+    this.container = container;
+  }
+
+  init() {
+    if (Site.currentProp("mobileView") || !("IntersectionObserver" in window)) {
+      return;
+    }
+
+    const appEvents = this.container.lookup("service:app-events");
+    appEvents.on("topic:current-post-scrolled", this._handlePostNodes);
+    appEvents.on("topic:scrolled", this._handleScroll);
+    appEvents.on("page:topic-loaded", this._initIntersectionObserver);
+
+    addWidgetCleanCallback("post-stream", this._clearIntersectionObserver);
+  }
+
+  @bind
+  _handleScroll(offset) {
+    if (offset <= 0) {
+      this.direction = "⬇️";
+      document
+        .querySelectorAll(`${this.topicPostSelector}.${this.stickyClass}`)
+        .forEach((node) => node.classList.remove(this.stickyClass));
+    } else if (offset > this.prevOffset) {
+      this.direction = "⬇️";
+    } else {
+      this.direction = "⬆️";
+    }
+    this.prevOffset = offset;
+  }
+
+  @bind
+  _handlePostNodes() {
+    this._clearIntersectionObserver();
+
+    schedule("afterRender", () => {
+      this._initIntersectionObserver();
+
+      document.querySelectorAll(this.topicPostSelector).forEach((postNode) => {
+        this.intersectionObserver.observe(postNode);
+
+        const topicAvatarNode = postNode.querySelector(".topic-avatar");
+        if (!topicAvatarNode || !postNode.querySelector("#post_1")) {
+          return;
+        }
+
+        const topicMapNode = postNode.querySelector(".topic-map");
+        if (!topicMapNode) {
+          return;
+        }
+        topicAvatarNode.style.marginBottom = `${topicMapNode.clientHeight}px`;
+      });
+    });
+  }
+
+  @bind
+  _initIntersectionObserver() {
+    const headerHeight = document.querySelector(".d-header")?.clientHeight || 0;
```

Comment:
It would safer to pull the same `--header-offset` value here I think. The header height calculation works fine in core, but some custom themes/plugins add other sticky elements above the header, and the header offset calculation already accounts for those cases.

Reply:
:+1:

Works for me.

You are talking of doing this `document.documentElement.style.getPropertyValue("--header-offset")` right?

I'm surprised we don't do this at any place in core yet though.

Reply:
Just pushed the change :+1: Had to use `parseInt` though as `getPropertyValue` returns a string in `px` and the intersectionObserver rootMargin property doesn’t seem to like non integer values :thinking: